### ISSUE TITLE
Add fairness 

### ISSUE THREAD

**Issue by pplonski** [mljar/mljar-supervised#612]
Implement AutoML fairness based on https://arxiv.org/abs/2111.06495 by @qingyun-wu and @sonichi

### Requirements:
- fairness training will be triggered when sensitive attributes are set in AutoML fit method,
- there should be several fairness metrics available with option to set custom metric, 
- fairness should be available in all four modes of MLJAR AutoML for all algorithms (including ensemble and stacked ensemble),
- automatic documentation should have information about fairness metrics and include comparison of models,
- provide unit tests to check new feature and performance tests,
- update docs,
- provide examples/tutorials.

### Example code

```python
X, y = load_training_data()

# init AutoML
automl = AutoML()

# case 1) training with sensitive attributes, use default fairness_metric
automl.fit(X, y, sensitive_features=["feature1", "feature2"])

# case 2) training with sensitive attributes and select fairness_metric
automl.fit(X, y, sensitive_features=["feature1", "feature2"], fairness_metric="equalized_odds")

# case 3) training with sensitive attributes and set custom fairness_metric
def custom fairness_metric(y_true, y_pred, sensitive_features, sample_weight=None):
    # implementation

automl.fit(X, y, sensitive_features=["feature1", "feature2"], fairness_metric=custom_fairness_metric)


```

**Comment by pplonski** [comment]
Hi @codeboy5!

Thank you for your offer! I started to study FairAutoML more, I don't like the approach from the paper. It looks good in theory, but for real-life problems might be unusable. Just imagine mitigating unfairness when doing 10-fold Cross-Validation. I think that applying Exponentiated Gradient to each model in each fold might be very inefficient. I found that Exponentiated Griadent has trouble optimizing for more than 1 sensitive feature, for example, if you have 2 sensitive features (A and B), then a mitigated model might be fair for feature A but unfair for feature B...

I would like to have a method that will search for a sample weight that will provide fairness. Then I would like to reuse the same sample weight when doing hyperparameters search.

So I'm in the process of searching for a method for fair-optimal same weighting... 

@codeboy5 do you have experience in fair ML or in optimization theory?

**Comment by pplonski** [comment]
I created a `fairness` module. It can compute fairness metrics and plots for binary classification tasks. It compute statistics for every sensitive feature separately.

Link to the module:
https://github.com/mljar/mljar-supervised/tree/fairness/supervised/fairness

Example script that compute fairness metrics:
https://github.com/mljar/mljar-supervised/blob/fairness/examples/scripts/binary_classifier_adult_fairness.py

The API:
```python
automl = AutoML(algorithms=["Xgboost"])
automl.fit(X_train, y_train, sensitive_features=sensitive_features_train)
```

TODO:
1. The preprocessing for sensitive features should be improved. We need to remove rows with missing sensitive feature value and we should remove sensitive feature rows when target is missing.
2. Only split validation supports sensitive features right now. It should be extended to all supported validation strategies.
3. Better handle situations when senstive features and sample weights are provided.

### Example report with information about fairness:

![fairness-metrics](https://user-images.githubusercontent.com/6959032/234578526-9d573202-e8c0-4f4d-b610-4c67491da255.gif)

**Comment by pplonski** [comment]
Improvements:
- added `fairness_metric`, `fariness_threshold` and `protected_groups` in the API,
- there are defaults set for `fairness_metric` and `fairness_threshold`,
- changed selection rate plot to vertical, added background for fair area, fair area is computed based on `fairness_threshold`
- changed the number format from % to 0.0000
- added information about fairness in the leaderboard - fairness is reported separately for each sensitive feature

Questions:
- maybe there should be `protected_groups` and `unprotected_groups`?


Demo:
![Peek 2023-05-01 13-34](https://user-images.githubusercontent.com/6959032/235446344-755ca39d-c0d6-4f82-8818-6f31f5e43c31.gif)

**Comment by pplonski** [comment]
I think there should be at least 20 samples of same group to be considered in fairness mitigation. For example, we have a group defined by "Female", "Young<30", "Black", and there are only 5 samples for this group (0 samples with class 1). This group shouldn't be considered for computing fairness metrics, and shouldn't be considered for fairness mitigation.

**Comment by pplonski** [comment]
I've pushed the `work in progress` version for fairness mitigation. There are a lot of prints in the terminal - it is working version.

The algorithm is optimizing demographic parity ratio (it is hard coded). The output for mitigation for single feature (`sex`):
![single-feature-mitigation](https://user-images.githubusercontent.com/6959032/235740396-55df7663-4dad-4392-a2d0-7b878f1b9328.gif)


The output for mitigation for two features (`sex`, `is_young`), `is_young` is categrocial feature craeted from `age<50`:
![two-features-mitigation](https://user-images.githubusercontent.com/6959032/235740432-99147812-ff9e-443b-824b-6a26bc23b980.gif)

**Comment by pplonski** [comment]
TODO:
- [ ] add automatic threhold setup
- [ ] add `privileged_groups` and `unprivileged_groups` in the API
- [ ] add information about privileged and unprivileged values
- [ ] automatically convert continuous sensitive features to binary 
- [ ] add green shadow for fair values and red for unfair values, add text annotations
- [ ] add fairness scatter plots for leaderboard

**Comment by pplonski** [comment]
If feature is not categorical it is automatically converted into binary feature based on equal samples number in each bin. We print the information in the terminal, example:
```
Sensitive features should be categorical
Apply automatic binarization for feature age
New values ['(37.0, 90.0]', '(16.999, 37.0]'] for feature age are applied
```

**Comment by pplonski** [comment]
The weights optimization stop condition is not yet implemented. This gives interesting behavior of algorithm. I was running the algorithm with `privileged_groups` and `unprivileged_groups` hard-coded and DP Ratio goes above 1.0.

Please notice that below script is using two sensitive features `sex` and `age`. The privileged group is defined only for `sex` feature, and for this feature ratio is going above 1.0 (because there is no stop condition).

The `age` is passed as continuous features that is automatically converted into binary.

```python
import pandas as pd
from sklearn.model_selection import train_test_split
from supervised.automl import AutoML
from sklearn.datasets import fetch_openml

data = fetch_openml(data_id=1590, as_frame=True)
y = (data.target == ">50K") * 1

X = data.data
y = (data.target == ">50K") * 1

sensitive_features = X[["sex", "age"]] 

X_train, X_test, y_train, y_test, S_train, S_test = train_test_split(
    X, y, sensitive_features, stratify=y, test_size=0.5, random_state=42
)

automl = AutoML(algorithms=["Xgboost"],
                train_ensemble=False,
                fairness_metric="demographic_parity_ratio",  # 
                fairness_threshold=0.8,
                privileged_groups = [{"sex": "Male"}],
                unprivileged_groups = [{"sex": "Female"}]
            )

automl.fit(X_train, y_train, sensitive_features=S_train)
```

Output:
![fairness-above-1 0](https://user-images.githubusercontent.com/6959032/236428493-879a32a1-6dc6-4e62-83f7-dd66c1590541.gif)

**Comment by pplonski** [comment]
I've run the AutoML with several algorithms on Adults dataset with two sensitive features. Below is an example output:
![Peek 2023-05-05 19-38](https://user-images.githubusercontent.com/6959032/236527854-929e20fa-47af-41c5-b353-7705fea4bf15.gif)

**Comment by pplonski** [comment]
### Preview of Fair Ensemble
![Peek 2023-05-17 13-09](https://github.com/mljar/mljar-supervised/assets/6959032/259b2905-d035-4939-838c-aeda91c5d51c)

**Comment by pplonski** [comment]
Issues:

 - [ ] nan values for fairness metrics for binary classification on categorical target
 - [ ] nan values for equalized odds metrics

**Comment by mosaikme** [comment]
Can we disable the Fairnes Metric?

**Comment by pplonski** [comment]
Hi @mosaikme, 

fairness is only used when you pass `sensitive_features` in `fit()` otherwise it is skipped.
